onnxruntime/core/providers/rocm/nn/pool.cc

```diff
@@ -0,0 +1,293 @@
+             const gsl::span<const int64_t>& kernel_shape,
+             const gsl::span<const int64_t>& pads,
+             const gsl::span<const int64_t>& strides) {
+    if (!desc_)
+      MIOPEN_RETURN_IF_ERROR(miopenCreatePoolingDescriptor(&desc_));
+
+    int rank = gsl::narrow_cast<int>(kernel_shape.size());
+    InlinedVector<int> window(rank);
+    InlinedVector<int> padding(rank);
+    InlinedVector<int> stride(rank);
+    for (int i = 0; i < rank; i++) {
+      window[i] = gsl::narrow_cast<int>(kernel_shape[i]);
+    }
+    for (int i = 0; i < rank; i++) {
+      padding[i] = gsl::narrow_cast<int>(pads[i]);
+    }
+    for (int i = 0; i < rank; i++) {
+      stride[i] = gsl::narrow_cast<int>(strides[i]);
+    }
+    MIOPEN_RETURN_IF_ERROR(miopenSetNdPoolingDescriptor(
+        desc_,
+        mode,
+        // MIOPEN_PROPAGATE_NAN,
+        rank,
+        window.data(),
+        padding.data(),
+        stride.data()));
+
+    return Status::OK();
+  }
+
+  operator miopenPoolingDescriptor_t() const { return desc_; }
+
+ private:
+  miopenPoolingDescriptor_t desc_;
+};
+
+template <typename T, typename PoolType>
+Status Pool<T, PoolType>::ComputeInternal(OpKernelContext* context) const {
+  typedef typename ToHipType<T>::MappedType HipT;
+  const Tensor* X = context->Input<Tensor>(0);
+  const TensorShape& x_shape = X->Shape();
+  const auto x_dims = x_shape.GetDims();
+
+  if (x_shape.NumDimensions() < 3) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Input dimension cannot be less than 3.");
+  }
+
+  auto kernel_shape = pool_attrs_.kernel_shape;
+  auto pads = pool_attrs_.pads;
+  auto strides = pool_attrs_.strides;
+
+  if (pool_attrs_.global_pooling) {
+    kernel_shape.assign(x_dims.begin() + 2, x_dims.end());
+    pads.assign(kernel_shape.size(), 0);
+    strides.assign(kernel_shape.size(), 1);
+  }
+
+  auto y_dims = pool_attrs_.SetOutputSize(x_shape, x_shape[1], &pads);
+  TensorShape y_shape(y_dims);
+  Tensor* Y = context->Output(0, y_shape);
+  // special case when there is a dim value of 0 in the shape.
+  if (y_shape.Size() == 0)
+    return Status::OK();
+
+  auto x_data = reinterpret_cast<const HipT*>(X->template Data<T>());
+  auto y_data = reinterpret_cast<HipT*>(Y->template MutableData<T>());
+
+  TensorShapeVector x_dims_miopen(x_dims.cbegin(), x_dims.cend());
+  TensorShapeVector y_dims_miopen(y_dims);
+  if (kernel_shape.size() < 2) {
+    // miopen only takes 4D or 5D input, so pad dimensions if needed
+    x_dims_miopen.push_back(1);
+    y_dims_miopen.push_back(1);
+    pads.insert(pads.begin() + kernel_shape.size(), 0);
+    pads.insert(pads.end(), 0);
+    kernel_shape.push_back(1);
+    strides.push_back(1);
+  }
+
+  miopenPoolingMode_t mode = miopenPoolingMax;
+  if constexpr (PoolType::type == onnxruntime::PoolType::kAveragePool) {
+    mode = pool_attrs_.count_include_pad ? miopenPoolingAverageInclusive
+                                         : miopenPoolingAverage;
```

Comment:
are these equivalent?     mode = pool_attrs_.count_include_pad ? CUDNN_POOLING_AVERAGE_COUNT_INCLUDE_PADDING
                                         : CUDNN_POOLING_AVERAGE_COUNT_EXCLUDE_PADDING;

Reply:
Also the same question for all related PRs - Is there any model requires this Op to run on AMD?

Reply:
> are these equivalent? mode = pool_attrs_.count_include_pad ? CUDNN_POOLING_AVERAGE_COUNT_INCLUDE_PADDING : CUDNN_POOLING_AVERAGE_COUNT_EXCLUDE_PADDING;

any further details on this comparison?

Reply:
https://github.com/ROCmSoftwarePlatform/hipDNN/blob/e7579a4ed8d38acf054c2a4c2796ce895d03cfcf/library/src/hcc_detail/hipdnn_miopen.cpp#L241-L246
They serve  the same purpose
CUDNN_POOLING_AVERAGE_COUNT_INCLUDE_PADDING -> (HIPIFY) HIPDNN_POOLING_AVERAGE_COUNT_INCLUDE_PADDING -> (hipDNN) miopenPoolingAverageInclusive